File: lang/fall/src/analysis/diagnostics.rs
use fall_tree::Node;

use fall_editor::{Diagnostic, Severity};

pub (crate) struct DiagnosticSink<'d> {
    diagnostics: &'d mut Vec<Diagnostic>
}

impl<'d> DiagnosticSink<'d> {
    pub fn new(diagnostics: &'d mut Vec<Diagnostic>) -> DiagnosticSink<'d> {
        DiagnosticSink { diagnostics }
    }

    pub fn error<T: Into<String>>(&mut self, node: Node, message: T) {
        self.diagnostics.push(Diagnostic {
            range: node.range(),
            severity: Severity::Error,
            message: message.into(),
        })
    }

    pub fn warning<T: Into<String>>(&mut self, node: Node, message: T) {
        self.diagnostics.push(Diagnostic {
            range: node.range(),
            severity: Severity::Warning,
            message: message.into(),
        })
    }
}

File: code/src/diagnostics.ts
import * as vscode from 'vscode'
import { FallDiagnostic} from './backend'
import { toVsRange } from './range'

const fallDiagnostics = vscode.languages.createDiagnosticCollection("fall")

export function setDiagnostics(
    document: vscode.TextDocument,
    diagnostics: Array<FallDiagnostic>
) {
    fallDiagnostics.clear()
    fallDiagnostics.set(
        document.uri,
        diagnostics.map((d) => {
            let range = toVsRange(document, d.range)
            let severity = d.severity == "Error"
                ? vscode.DiagnosticSeverity.Error
                : vscode.DiagnosticSeverity.Warning

            return new vscode.Diagnostic(range, d.message, severity)
        })
    )
}
